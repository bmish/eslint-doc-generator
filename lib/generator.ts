import { EOL } from 'node:os';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join, relative, resolve } from 'node:path';
import { getAllNamedOptions, hasOptions } from './rule-options.js';
import {
  loadPlugin,
  getPluginPrefix,
  getPluginRoot,
  getPathWithExactFileNameCasing,
} from './package-json.js';
import { updateRulesList } from './rule-list.js';
import { updateConfigsList } from './config-list.js';
import { generateRuleHeaderLines } from './rule-doc-notices.js';
import {
  parseRuleDocNoticesOption,
  parseRuleListColumnsOption,
  parseConfigEmojiOptions,
} from './option-parsers.js';
import {
  BEGIN_RULE_OPTIONS_LIST_MARKER,
  END_RULE_HEADER_MARKER,
  END_RULE_OPTIONS_LIST_MARKER,
} from './comment-markers.js';
import {
  replaceOrCreateHeader,
  expectContentOrFail,
  expectSectionHeaderOrFail,
} from './markdown.js';
import { resolveConfigsToRules } from './plugin-config-resolution.js';
import { OPTION_DEFAULTS } from './options.js';
import { diff } from 'jest-diff';
import type { GenerateOptions } from './types.js';
import { OPTION_TYPE, RuleModule } from './types.js';
import { replaceRulePlaceholder } from './rule-link.js';
import { updateRuleOptionsList } from './rule-options-list.js';

function stringOrArrayWithFallback<T extends string | readonly string[]>(
  stringOrArray: undefined | T,
  fallback: T
): T {
  return stringOrArray && stringOrArray.length > 0 ? stringOrArray : fallback;
}

function stringOrArrayToArrayWithFallback(
  stringOrArray: undefined | string | readonly string[],
  fallback: readonly string[]
): readonly string[] {
  const asArray =
    stringOrArray instanceof Array // eslint-disable-line unicorn/no-instanceof-array -- using Array.isArray() loses type information about the array.
      ? stringOrArray
      : stringOrArray
      ? [stringOrArray]
      : [];
  const csvStringItem = asArray.find((item) => item.includes(','));
  if (csvStringItem) {
    throw new Error(
      `Provide property as array, not a CSV string: ${csvStringItem}`
    );
  }
  return asArray && asArray.length > 0 ? asArray : fallback;
}

// eslint-disable-next-line complexity
export async function generate(path: string, options?: GenerateOptions) {
  const plugin = await loadPlugin(path);
  const pluginPrefix = getPluginPrefix(path);
  const configsToRules = await resolveConfigsToRules(plugin);

  if (!plugin.rules) {
    throw new Error('Could not find exported `rules` object in ESLint plugin.');
  }

  // Options. Add default values as needed.
  const check = options?.check ?? OPTION_DEFAULTS[OPTION_TYPE.CHECK];
  const configEmojis = parseConfigEmojiOptions(plugin, options?.configEmoji);
  const configFormat =
    options?.configFormat ?? OPTION_DEFAULTS[OPTION_TYPE.CONFIG_FORMAT];
  const ignoreConfig = stringOrArrayWithFallback(
    options?.ignoreConfig,
    OPTION_DEFAULTS[OPTION_TYPE.IGNORE_CONFIG]
  );
  const ignoreDeprecatedRules =
    options?.ignoreDeprecatedRules ??
    OPTION_DEFAULTS[OPTION_TYPE.IGNORE_DEPRECATED_RULES];
  const initRuleDocs =
    options?.initRuleDocs ?? OPTION_DEFAULTS[OPTION_TYPE.INIT_RULE_DOCS];
  const pathRuleDoc =
    options?.pathRuleDoc ?? OPTION_DEFAULTS[OPTION_TYPE.PATH_RULE_DOC];
  const pathRuleList = stringOrArrayToArrayWithFallback(
    options?.pathRuleList,
    OPTION_DEFAULTS[OPTION_TYPE.PATH_RULE_LIST]
  );
  const postprocess =
    options?.postprocess ?? OPTION_DEFAULTS[OPTION_TYPE.POSTPROCESS];
  const ruleDocNotices = parseRuleDocNoticesOption(options?.ruleDocNotices);
  const ruleDocSectionExclude = stringOrArrayWithFallback(
    options?.ruleDocSectionExclude,
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_SECTION_EXCLUDE]
  );
  const ruleDocSectionInclude = stringOrArrayWithFallback(
    options?.ruleDocSectionInclude,
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_SECTION_INCLUDE]
  );
  const ruleDocSectionOptions =
    options?.ruleDocSectionOptions ??
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_SECTION_OPTIONS];
  const ruleDocTitleFormat =
    options?.ruleDocTitleFormat ??
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_TITLE_FORMAT];
  const ruleListColumns = parseRuleListColumnsOption(options?.ruleListColumns);
  const ruleListSplit =
    typeof options?.ruleListSplit === 'function'
      ? options.ruleListSplit
      : stringOrArrayToArrayWithFallback(
          options?.ruleListSplit,
          OPTION_DEFAULTS[OPTION_TYPE.RULE_LIST_SPLIT]
        );
  const urlConfigs =
    options?.urlConfigs ?? OPTION_DEFAULTS[OPTION_TYPE.URL_CONFIGS];
  const urlRuleDoc =
    options?.urlRuleDoc ?? OPTION_DEFAULTS[OPTION_TYPE.URL_RULE_DOC];

  // Gather normalized list of rules.
  const ruleNamesAndRules = Object.entries(plugin.rules)
    .map(([name, ruleModule]) => {
      // Convert deprecated function-style rules to object-style rules so that we don't have to handle function-style rules everywhere throughout the codebase.
      // @ts-expect-error -- this type unfortunately requires us to choose a `meta.type` even though the deprecated function-style rule won't have one.
      const ruleModuleAsObject: RuleModule =
        typeof ruleModule === 'function'
          ? {
              // Deprecated function-style rule don't support most of the properties that object-style rules support, so we'll just use the bare minimum.
              meta: {
                // @ts-expect-error -- type is missing for this property
                schema: ruleModule.schema, // eslint-disable-line @typescript-eslint/no-unsafe-assignment -- type is missing for this property
                // @ts-expect-error -- type is missing for this property
                deprecated: ruleModule.deprecated, // eslint-disable-line @typescript-eslint/no-unsafe-assignment -- type is missing for this property
              },
              create: ruleModule,
            }
          : ruleModule;
      const tuple: [string, RuleModule] = [name, ruleModuleAsObject];
      return tuple;
    })
    .filter(
      // Filter out deprecated rules from being checked, displayed, or updated if the option is set.
      ([, rule]) => !ignoreDeprecatedRules || !rule.meta.deprecated
    )
    .sort(([a], [b]) => a.toLowerCase().localeCompare(b.toLowerCase()));

  // Update rule doc for each rule.
  let initializedRuleDoc = false;
  for (const [name, rule] of ruleNamesAndRules) {
    const schema = rule.meta?.schema;
    const description = rule.meta?.docs?.description;
    const pathToDoc = join(path, replaceRulePlaceholder(pathRuleDoc, name));
    const ruleHasOptions = hasOptions(schema);

    if (!existsSync(pathToDoc)) {
      if (!initRuleDocs) {
        throw new Error(
          `Could not find rule doc (run with --init-rule-docs to create): ${relative(
            getPluginRoot(path),
            pathToDoc
          )}`
        );
      }

      // Determine content for fresh rule doc, including any mandatory sections.
      // The rule doc header will be added later.
      let newRuleDocContents = [
        ruleDocSectionInclude.length > 0
          ? ruleDocSectionInclude
              .map((title) => `## ${title}`)
              .join(`${EOL}${EOL}`)
          : undefined,
        ruleHasOptions
          ? `## Options${EOL}${EOL}${BEGIN_RULE_OPTIONS_LIST_MARKER}${EOL}${END_RULE_OPTIONS_LIST_MARKER}`
          : undefined,
      ]
        .filter((section) => section !== undefined)
        .join(`${EOL}${EOL}`);
      if (newRuleDocContents !== '') {
        newRuleDocContents = `${EOL}${newRuleDocContents}${EOL}`;
      }

      mkdirSync(dirname(pathToDoc), { recursive: true });
      writeFileSync(pathToDoc, newRuleDocContents);
      initializedRuleDoc = true;
    }

    // Regenerate the header (title/notices) of each rule doc.
    const newHeaderLines = generateRuleHeaderLines(
      description,
      name,
      plugin,
      configsToRules,
      pluginPrefix,
      path,
      pathRuleDoc,
      configEmojis,
      configFormat,
      ignoreConfig,
      ruleDocNotices,
      ruleDocTitleFormat,
      urlConfigs,
      urlRuleDoc
    );

    const contentsOld = readFileSync(pathToDoc).toString();
    const contentsNew = await postprocess(
      updateRuleOptionsList(
        replaceOrCreateHeader(
          contentsOld,
          newHeaderLines,
          END_RULE_HEADER_MARKER
        ),
        rule
      ),
      resolve(pathToDoc)
    );

    if (check) {
      if (contentsNew !== contentsOld) {
        console.error(
          `Please run eslint-doc-generator. A rule doc is out-of-date: ${relative(
            getPluginRoot(path),
            pathToDoc
          )}`
        );
        console.error(diff(contentsNew, contentsOld, { expand: false }));
        process.exitCode = 1;
      }
    } else {
      writeFileSync(pathToDoc, contentsNew);
    }

    // Check for potential issues with the rule doc.

    // Check for required sections.
    for (const section of ruleDocSectionInclude) {
      expectSectionHeaderOrFail(
        `\`${name}\` rule doc`,
        contentsNew,
        [section],
        true
      );
    }

    // Check for disallowed sections.
    for (const section of ruleDocSectionExclude) {
      expectSectionHeaderOrFail(
        `\`${name}\` rule doc`,
        contentsNew,
        [section],
        false
      );
    }

    if (ruleDocSectionOptions) {
      // Options section.
      expectSectionHeaderOrFail(
        `\`${name}\` rule doc`,
        contentsNew,
        ['Options', 'Config'],
        ruleHasOptions
      );
      for (const { name: namedOption } of getAllNamedOptions(schema)) {
        expectContentOrFail(
          `\`${name}\` rule doc`,
          'rule option',
          contentsNew,
          namedOption,
          true
        ); // Each rule option is mentioned.
      }
    }
  }

  if (initRuleDocs && !initializedRuleDoc) {
    throw new Error(
      '--init-rule-docs was enabled, but no rule doc file needed to be created.'
    );
  }

  for (const pathRuleListItem of pathRuleList) {
    // Find the exact filename.
    const pathToFile = getPathWithExactFileNameCasing(
      join(path, pathRuleListItem)
    );
    if (!pathToFile || !existsSync(pathToFile)) {
      throw new Error(
        `Could not find ${String(pathRuleList)} in ESLint plugin.`
      );
    }

    // Update the rules list in this file.
    const fileContents = readFileSync(pathToFile, 'utf8');
    const fileContentsNew = await postprocess(
      updateConfigsList(
        updateRulesList(
          ruleNamesAndRules,
          fileContents,
          plugin,
          configsToRules,
          pluginPrefix,
          pathRuleDoc,
          pathToFile,
          path,
          configEmojis,
          configFormat,
          ignoreConfig,
          ruleListColumns,
          ruleListSplit,
          urlConfigs,
          urlRuleDoc
        ),
        plugin,
        configsToRules,
        pluginPrefix,
        configEmojis,
        configFormat,
        ignoreConfig
      ),
      resolve(pathToFile)
    );

    if (check) {
      if (fileContentsNew !== fileContents) {
        console.error(
          `Please run eslint-doc-generator. The rules table in ${relative(
            getPluginRoot(path),
            pathToFile
          )} is out-of-date.`
        );
        console.error(diff(fileContentsNew, fileContents, { expand: false }));
        process.exitCode = 1;
      }
    } else {
      writeFileSync(pathToFile, fileContentsNew, 'utf8');
    }
  }
}
