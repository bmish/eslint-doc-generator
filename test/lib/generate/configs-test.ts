import { generate } from '../../../lib/generator.js';
import mockFs from 'mock-fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { readFileSync } from 'node:fs';
import { jest } from '@jest/globals';

const __dirname = dirname(fileURLToPath(import.meta.url));

const PATH_NODE_MODULES = resolve(__dirname, '..', '..', '..', 'node_modules');

describe('generate (configs)', function () {
  describe('config with overrides', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': {
                    meta: { docs: { description: 'Description of no-foo.' }, },
                    create(context) {}
                  },
                },
                configs: {
                  recommended: {
                    overrides: [{
                      files: ['**/foo.js'],
                      rules: {
                        'test/no-foo': 'error',
                      }
                    }]
                  },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('generates the documentation', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('rule config with options', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': {
                    meta: { docs: { description: 'Description of no-foo.' }, },
                    create(context) {},
                    schema: [{ /* some options */ }]
                  },
                },
                configs: {
                  recommended: {
                    rules: {
                      'test/no-foo': ['error', { /* some options */ }],
                    }
                  },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('generates the documentation', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('rules that are disabled or set to warn', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': {
                    meta: { docs: { description: 'Description of no-foo.' }, },
                    create(context) {},
                  },
                  'no-bar': {
                    meta: { docs: { description: 'Description of no-bar.' }, },
                    create(context) {},
                  },
                  'no-baz': {
                    meta: { docs: { description: 'Description of no-baz.' }, },
                    create(context) {},
                  },
                  'no-biz': {
                    meta: { docs: { description: 'Description of no-biz.' }, },
                    create(context) {},
                  },
                  'no-boz': {
                    meta: { docs: { description: 'Description of no-boz.' }, },
                    create(context) {},
                  },
                  'no-buz': {
                    meta: { docs: { description: 'Description of no-buz.' }, },
                    create(context) {},
                  },
                  'no-bez': {
                    meta: { docs: { description: 'Description of no-bez.' }, },
                    create(context) {},
                  },
                },
                configs: {
                  recommended: {
                    rules: {
                      'test/no-foo': 'off',
                      'test/no-bar': 0,
                      'test/no-baz': 'error',
                      'test/no-boz': 'warn',
                      'test/no-buz': 1,
                    }
                  },
                  other: {
                    rules: {
                      'test/no-bar': 0,
                      'test/no-baz': 'off',
                      'test/no-biz': 'off',
                      'test/no-buz': 'warn',
                      'test/no-bez': 'warn',
                    }
                  },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',
        'docs/rules/no-bar.md': '',
        'docs/rules/no-baz.md': '',
        'docs/rules/no-biz.md': '',
        'docs/rules/no-boz.md': '',
        'docs/rules/no-buz.md': '',
        'docs/rules/no-bez.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('generates the documentation', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-bar.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-baz.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-biz.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-boz.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-buz.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-bez.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('rules that are disabled or set to warn, only one config present', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': {
                    meta: { docs: { description: 'Description of no-foo.' }, },
                    create(context) {},
                  },
                  'no-bar': {
                    meta: { docs: { description: 'Description of no-bar.' }, },
                    create(context) {},
                  },
                },
                configs: {
                  recommended: {
                    rules: {
                      'test/no-foo': 1,
                      'test/no-bar': 0,
                    }
                  },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',
        'docs/rules/no-bar.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('generates the documentation', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-bar.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('rules that are disabled or set to warn, two configs present', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': {
                    meta: { docs: { description: 'Description of no-foo.' }, },
                    create(context) {},
                  },
                },
                configs: {
                  recommended: {
                    rules: {
                      'test/no-foo': 1,
                    }
                  },
                  typescript: {
                    rules: {
                      'test/no-foo': 0,
                    }
                  },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('generates the documentation', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('with config that does not have any rules', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
          export default {
            rules: {
              'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
            },
            configs: {
              recommended: { rules: { 'test/no-foo': 'error' } },
              configWithoutRules: { rules: {  } },
            }
          };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('uses recommended config emoji since it is the only relevant config', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('with external rules in config', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
          export default {
            rules: {
              'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
            },
            configs: {
              recommended: {
                plugins: ['external', 'test'],
                rules: {
                  'test/no-foo': 'error',
                  'external/no-bar': 'error',
                }
              },
            }
          };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('ignores external rules', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('only a `recommended` config', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
          export default {
            rules: { 'no-foo': { meta: { docs: { description: 'Description.' }, }, create(context) {} }, },
            configs: {
              recommended: {
                rules: {
                  'test/no-foo': 'error',
                }
              }
            }
          };`,

        'README.md':
          '<!-- begin auto-generated rules list --><!-- end auto-generated rules list -->',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('updates the documentation', async function () {
      await generate('.');

      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();

      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('with --ignore-config', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
          export default {
            rules: {
              'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
            },
            configs: {
              recommended: {
                rules: { 'test/no-foo': 'error' },
              },
              configToIgnore: {
                rules: { 'test/no-foo': 'error' },
              }
            }
          };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('hides the ignored config', async function () {
      await generate('.', {
        ignoreConfig: ['configToIgnore'],
        configEmoji: [['configToIgnore', '😋']], // Ensure this config has an emoji that would normally display in the legend.
      });
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });
});
