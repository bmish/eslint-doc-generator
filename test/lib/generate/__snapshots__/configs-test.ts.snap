// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate (configs) config with overrides generates the documentation 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

💼 Configurations enabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description            | 💼 |
| :----------------------------- | :--------------------- | :- |
| [no-foo](docs/rules/no-foo.md) | Description of no-foo. | ✅  |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (configs) config with overrides generates the documentation 2`] = `
"# Description of no-foo (\`test/no-foo\`)

💼 This rule is enabled in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) only a \`recommended\` config updates the documentation 1`] = `
"<!-- begin auto-generated rules list -->

💼 Configurations enabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description  | 💼 |
| :----------------------------- | :----------- | :- |
| [no-foo](docs/rules/no-foo.md) | Description. | ✅  |

<!-- end auto-generated rules list -->"
`;

exports[`generate (configs) only a \`recommended\` config updates the documentation 2`] = `
"# Description (\`test/no-foo\`)

💼 This rule is enabled in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rule config with options generates the documentation 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

💼 Configurations enabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description            | 💼 |
| :----------------------------- | :--------------------- | :- |
| [no-foo](docs/rules/no-foo.md) | Description of no-foo. | ✅  |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (configs) rule config with options generates the documentation 2`] = `
"# Description of no-foo (\`test/no-foo\`)

💼 This rule is enabled in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

💼 Configurations enabled in.\\
⚠️ Configurations set to warn in.\\
🚫 Configurations disabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description            | 💼 | ⚠️                 | 🚫                 |
| :----------------------------- | :--------------------- | :- | :----------------- | :----------------- |
| [no-bar](docs/rules/no-bar.md) | Description of no-bar. |    |                    | ✅ ![badge-other][] |
| [no-baz](docs/rules/no-baz.md) | Description of no-baz. | ✅  |                    | ![badge-other][]   |
| [no-bez](docs/rules/no-bez.md) | Description of no-bez. |    | ![badge-other][]   |                    |
| [no-biz](docs/rules/no-biz.md) | Description of no-biz. |    |                    | ![badge-other][]   |
| [no-boz](docs/rules/no-boz.md) | Description of no-boz. |    | ✅                  |                    |
| [no-buz](docs/rules/no-buz.md) | Description of no-buz. |    | ✅ ![badge-other][] |                    |
| [no-foo](docs/rules/no-foo.md) | Description of no-foo. |    |                    | ✅                  |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 2`] = `
"# Description of no-foo (\`test/no-foo\`)

🚫 This rule is _disabled_ in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 3`] = `
"# Description of no-bar (\`test/no-bar\`)

🚫 This rule is _disabled_ in the following configs: \`other\`, ✅ \`recommended\`.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 4`] = `
"# Description of no-baz (\`test/no-baz\`)

💼🚫 This rule is enabled in the ✅ \`recommended\` config. This rule is _disabled_ in the \`other\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 5`] = `
"# Description of no-biz (\`test/no-biz\`)

🚫 This rule is _disabled_ in the \`other\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 6`] = `
"# Description of no-boz (\`test/no-boz\`)

⚠️ This rule _warns_ in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 7`] = `
"# Description of no-buz (\`test/no-buz\`)

⚠️ This rule _warns_ in the following configs: \`other\`, ✅ \`recommended\`.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn generates the documentation 8`] = `
"# Description of no-bez (\`test/no-bez\`)

⚠️ This rule _warns_ in the \`other\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn, only one config present generates the documentation 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

⚠️ Configurations set to warn in.\\
🚫 Configurations disabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description            | ⚠️ | 🚫 |
| :----------------------------- | :--------------------- | :- | :- |
| [no-bar](docs/rules/no-bar.md) | Description of no-bar. |    | ✅  |
| [no-foo](docs/rules/no-foo.md) | Description of no-foo. | ✅  |    |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn, only one config present generates the documentation 2`] = `
"# Description of no-foo (\`test/no-foo\`)

⚠️ This rule _warns_ in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn, only one config present generates the documentation 3`] = `
"# Description of no-bar (\`test/no-bar\`)

🚫 This rule is _disabled_ in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn, two configs present generates the documentation 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

⚠️ Configurations set to warn in.\\
🚫 Configurations disabled in.\\
✅ Set in the \`recommended\` configuration.\\
⌨️ Set in the \`typescript\` configuration.

| Name                           | Description            | ⚠️ | 🚫 |
| :----------------------------- | :--------------------- | :- | :- |
| [no-foo](docs/rules/no-foo.md) | Description of no-foo. | ✅  | ⌨️ |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (configs) rules that are disabled or set to warn, two configs present generates the documentation 2`] = `
"# Description of no-foo (\`test/no-foo\`)

⚠️🚫 This rule _warns_ in the ✅ \`recommended\` config. This rule is _disabled_ in the ⌨️ \`typescript\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) with --ignore-config hides the ignored config 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

💼 Configurations enabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description             | 💼 |
| :----------------------------- | :---------------------- | :- |
| [no-foo](docs/rules/no-foo.md) | Description for no-foo. | ✅  |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (configs) with --ignore-config hides the ignored config 2`] = `
"# Description for no-foo (\`test/no-foo\`)

💼 This rule is enabled in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) with config that does not have any rules uses recommended config emoji since it is the only relevant config 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

💼 Configurations enabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description             | 💼 |
| :----------------------------- | :---------------------- | :- |
| [no-foo](docs/rules/no-foo.md) | Description for no-foo. | ✅  |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (configs) with config that does not have any rules uses recommended config emoji since it is the only relevant config 2`] = `
"# Description for no-foo (\`test/no-foo\`)

💼 This rule is enabled in the ✅ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (configs) with external rules in config ignores external rules 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

💼 Configurations enabled in.\\
✅ Set in the \`recommended\` configuration.

| Name                           | Description             | 💼 |
| :----------------------------- | :---------------------- | :- |
| [no-foo](docs/rules/no-foo.md) | Description for no-foo. | ✅  |

<!-- end auto-generated rules list -->
"
`;
